lib/forwarder/handler.go

```diff
@@ -0,0 +1,210 @@
+package forwarder
+
+import (
+	"context"
+	"runtime/debug"
+	"tcplb/lib/core"
+	"tcplb/lib/limiter"
+	"tcplb/lib/slog"
+)
+
+const (
+	RecovererUnexpectedPanicMessage = "RecovererHandler: Unexpected panic!"
+)
+
+type clientIdContextKeyType struct{}
+type upstreamsContextKeyType struct{}
+
+var clientIdContextKey = clientIdContextKeyType{}
+var upstreamContextKey = upstreamsContextKeyType{}
+
+func NewContextWithClientID(parent context.Context, clientID core.ClientID) context.Context {
+	return context.WithValue(parent, clientIdContextKey, clientID)
+}
+
+func ClientIDFromContext(ctx context.Context) (core.ClientID, bool) {
+	clientID, ok := ctx.Value(clientIdContextKey).(core.ClientID)
+	return clientID, ok
+}
+
+func NewContextWithUpstreams(parent context.Context, upstreams core.UpstreamSet) context.Context {
+	return context.WithValue(parent, upstreamContextKey, upstreams)
+}
+
+func UpstreamsFromContext(ctx context.Context) (core.UpstreamSet, bool) {
+	upstreams, ok := ctx.Value(upstreamContextKey).(core.UpstreamSet)
+	return upstreams, ok
+}
+
+type Handler interface {
+	// Handle accepts the given AuthenticatedConn from the client.
+	Handle(ctx context.Context, conn AuthenticatedConn)
+}
+
+// ConnCloserHandler is a handler that closes the client connection
+// after the Inner handler has finished handling it. It should be the
+// base Handler in the stack.
+type ConnCloserHandler struct {
+	Inner Handler
+}
+
+func (h *ConnCloserHandler) Handle(ctx context.Context, conn AuthenticatedConn) {
+	defer func() {
+		// If there are errors closing the client connection, it is
+		// likely due to client or network. Ignore them.
+		_ = conn.Close()
+	}()
+	h.Inner.Handle(ctx, conn)
+}
+
+var _ Handler = (*ConnCloserHandler)(nil) // type check
+
+// RecovererHandler is a handler that recovers from panics raised by the
+// Inner handler (if any) and logs an error to the given Logger.
+type RecovererHandler struct {
+	Logger slog.Logger
+	Inner  Handler
+}
+
+func (h *RecovererHandler) Handle(ctx context.Context, conn AuthenticatedConn) {
+	defer func() {
+		panicValue := recover()
```

Comment:
This recover makes me uneasy, recovering is not typically something you want to do. I wouldn't expect to spot a recover in the typical challenge program.

Reply:
I hold two conflicting perspectives about this kind of thing:

suppose some part of the application code can panic in an unexpected way that wasn't caught by unit testing, review, static analysis, load testing etc. If the defect can cause a panic in some scenarios, then an unhandled panic in a single goroutine causes the entire program to terminate.

so one perspective is: assuming the defect is limited to a single goroutine, let the stack unwind, then trap the panic at the end, log it, let that goroutine terminate, but keep on running.  this perspective assumes that the cause of the panic is likely "local" to the goroutine or the client connection it was handling, so trapping the panic reduces the blast radius of the unexpected fault - it reduces the chance we also impact useful work that other goroutines are doing.

another perspective is: we do not expect our program to panic, it is not a part of the intended control flow, if it panics it is definitely a defect, and further it might be a symptom that the program has entered into dangerously undefined state (maybe there's memory corruption impacting multiple goroutines, not just the one that happens to have panicked). from this perspective it is reasonable to regard the entire program as suspect, and terminate it.

in dev & test & QA it makes a lot of sense to bias toward maximising the chance of detecting defects, so they can be removed.  so we should remove the recover here, as the goal is to write low or zero defect code, and shake the defects out.  i'll turn it off.

but in a real production situation, we might want something like this turned on, or some other mechanism in our design to stop local failures caused by the defects that sometimes ship turning into global unavailability.

Reply:
it is removed